Source/SwiftLintFramework/Rules/Style/NoGuardReturnVoidRule.swift

```diff
@@ -0,0 +1,196 @@
+                }
+                differentSideEffect()
+            }
+            """,
+            """
+            func test() {
+                guard otherCondition else {
+                    return↓ assertionfailure(""); // comment
+                }
+                differentSideEffect()
+            }
+            """
+        ]
+    )
+
+    public func validate(file: File, kind: SwiftDeclarationKind,
+                         dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+        guard SwiftDeclarationKind.functionKinds.contains(kind),
+            dictionary.isVoid(),
+            !dictionary.isInit(),
+            let guardRanges = dictionary.guardStatementRanges(),
+            !guardRanges.isEmpty
+        else {
+            return []
+        }
+
+        return guardRanges
+            .map({ range in
+                return file.syntaxMap.tokens.filter { range.contains($0.offset) && !$0.isComment }
+            })
+            .compactMap({ tokens in
+                guard let last = tokens.last,
+                    let lastReturn = tokens.last(where: { $0.isReturnKeyword(in: file) }),
+                    last != lastReturn
+                else {
+                    return nil
+                }
+
+                let location = Location(file: file, byteOffset: lastReturn.offset + lastReturn.length)
+                return StyleViolation(ruleDescription: type(of: self).description,
+                                      severity: configuration.severity,
+                                      location: location)
+            })
+    }
+}
+
+private extension Dictionary where Dictionary == [String: SourceKitRepresentable] {
+    func isVoid() -> Bool {
+        return self["key.typename"] == nil || (self["key.typename"] as? String) == "Void"
+    }
+
+    func isInit() -> Bool {
+        return (self["key.name"] as? String)?.hasPrefix("init(") ?? false
+    }
+
+    func guardStatementRanges() -> [NSRange]? {
+        let ranges = (self["key.substructure"] as? [[String: SourceKitRepresentable]])?
+            .compactMap({ $0 })
+            .filter({ $0.isGuardStatment() })
+            .compactMap({ $0.byteRange() })
+
+        return ranges
+    }
+
+    func isGuardStatment() -> Bool {
+        return self["key.kind"] as? String == "source.lang.swift.stmt.guard"
+    }
+
+    func byteRange() -> NSRange? {
+        guard let offset = self["key.offset"] as? Int64, let length = self["key.length"] as? Int64 else {
+            return nil
+        }
+
+        return NSRange(location: Int(offset), length: Int(length))
+    }
+}
+
+private extension SyntaxToken {
+    var isComment: Bool {
+        return self.type == "source.lang.swift.syntaxtype.comment"
+    }
+
+    var isKeyword: Bool {
+        return self.type == "source.lang.swift.syntaxtype.keyword"
+    }
+
+    func isReturnKeyword(in file: File) -> Bool {
+        return self.isKeyword && (file.contents(for: self) == "return")
+    }
+}
+
```

Comment:
Oops, you now have _two_ trailing newlines here. Unix standard is a single one: https://unix.stackexchange.com/a/18789/243788

Reply:
Ah, I was overshooting instead of undershooting. 🙄Thanks!